### PR TITLE
Documentation for using ConnectionPatch across Axes with constrained_…

### PR DESCRIPTION
**Pull request by canismarko**
…layout.

When using a figure with contrained layout management, adding a
ConnectionPatch between different subplot Axes breaks the layout. The
artist should be removed from layout considerations before being added
to the axes. This commit notes this fix in the documentation.

https://github.com/matplotlib/matplotlib/issues/14907

## PR Checklist

- [ ] Has Pytest style unit tests
- [ ] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant
- [ ] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

### CODE REVIEW

File: tutorials/intermediate/constrainedlayout_guide.py [hunk 1]
```diff
@@ -581,6 +581,12 @@ def docomplicated(suptitle=None):
 #
 # * There are small differences in how the backends handle rendering fonts,
 #   so the results will not be pixel-identical.
+#
+# * A :class:`~matplotlib.patches.ConnectionPatch` across two
+#   :class:`~matplotlib.axes.Axes` must be removed from the layout
+#   using :meth:`~matplotlib.artist.Artist.set_in_layout` before
+#   calling :meth:`~matplotlib.axes.Axes.add_artist`. See
+#   :class:`~matplotlib.patches.ConnectionPatch` for more information.
```
Comment (ImportanceOfBeingErnest) [inline tutorials/intermediate/constrainedlayout_guide.py]: In principle this applies to just any artist that uses axes coordinates and is added to an axes. So similar problems are expected for annotations or annotation boxes in axes coordinates, ax{h,v}lines if they overflow the axes boundary, legends (which are actually mentionned someplace earlier in this tutorial), etc. 

Maybe there is a way to formulate this more generally and name the connection patch as an example?

Comment (canismarko) [reply tutorials/intermediate/constrainedlayout_guide.py]: Yeah sure. I updated the constrained_layout tutorial to apply to all artists. Let me know what you think.

Comment (ImportanceOfBeingErnest) [reply tutorials/intermediate/constrainedlayout_guide.py]: Contentwise I'd be fine with that. I'll leave it to a native speaker to check if that sentence makes sense grammatically. Probably, @jklymak, the original author of that tutorial would like to have a look at it anyways.

**Comment by jklymak** [comment]
I am away from computer for a few days.  If folks think this is an improvement don’t wait for me!

**Comment by ImportanceOfBeingErnest** [comment]
Or maybe one should advocate the use of `fig.add_artist()` more in this respect? (Artists added to figures do not generally take part in the layout)  
That would include:

* mention it in the tight layout and constrained layout tutorial
* mention it in the docstring for artists that are often positionned outside axes, like ConnectionPatch, but also AnnotationBbox, `Annotation`
* use it in the [Connect Simple01](https://matplotlib.org/gallery/userdemo/connect_simple01.html) example. Also, the [Demo Annotation Box](https://matplotlib.org/gallery/text_labels_and_annotations/demo_annotation_box.html)

**Comment by anntzer** [comment]
I agree that using Figure.add_artist makes more sense than adding to the Axes and then excluding from the Axes layout...

**Comment by canismarko** [comment]
Alright, I will update the PR to suggest Figure.add_artist instead. Thanks.

**Comment by ImportanceOfBeingErnest** [comment]
Except that `fig.add_artist(ConnectionPatch(..))` doesn't currently work. So one would need to make it work first.

**Comment by ImportanceOfBeingErnest** [comment]
#15020 would make it work. Once it's merged documentation can use it.

**Comment by canismarko** [comment]
Since https://github.com/matplotlib/matplotlib/pull/15020 was merged a few days ago, I revised the PR to recommend using Figure.add_artist().

**Comment by tacaswell** [comment]
Looks good to me, leaving to @ImportanceOfBeingErnest to merge.

**Comment by tacaswell** [comment]
Thank you @canismarko and congratulations on your first Matplotlib PR :tada: Hopefully we will hear from you again!

**Comment by canismarko** [comment]
Thanks, @tacaswell. I'll see if I can implement the rest of the suggestions from @ImportanceOfBeingErnest in https://github.com/matplotlib/matplotlib/pull/14957#issuecomment-518014352 and open a new PR when they're done.

**Pull request merged**